examples/nn.dx

```diff
@@ -0,0 +1,227 @@
+' # Neural Networks
+
+include "plot.dx"
+
+' ## NN Prelude
+
+
+def relu (input : Float) : Float =
+  select (input > 0.0) input 0.0
+
+instance [Add a, Add b] Add (a & b)
+  add = \(a, b) (c, d). ( (a + c), (b + d))
+  sub = \(a, b) (c, d). ( (a - c), (b - d))
+  zero = (zero, zero)
+
+instance [VSpace a, VSpace b] VSpace (a & b)
+  scaleVec = \ s (a, b) . (scaleVec s a, scaleVec s b)
+
+def Layer (inp:Type) (out:Type) (param:Type) : Type=
+  {forward:(param -> inp -> out) &
+   init: Key -> param
+  }
+
+def forward (l:Layer i o p) (p : p) (x : i): o =
+  (getAt #forward l) p x
+
+def init (l:Layer i o p) (k:Key)  : p  =
+  (getAt #init l) k
+
+
+' ## Layers 
+
+' Dense layer
+
+def Dense (a:Type) (b:Type) : Type =
+  Layer (a=>Float) (b=>Float) ((a=>b=>Float) & (b=>Float))
+
+def dense (a:Type) (b:Type) : Dense a b =
+  {
+   forward = (\(weight, bias) x .
+               for j:b. (bias.j + sum for i :a. weight.i.j * x.i)),
+   init = arb
+  }
+
+' CNN layer
+
+def CNN (h:Int) (w:Int) (inc:Type) (outc:Type) (kw:Int) (kh:Int) : Type =
+  (Layer (inc=>Fin h=>Fin w=>Float)
+         (outc=>Fin h=>Fin w=>Float)
+         ((outc=>inc=>Fin kh=>Fin kw=>Float) &
+          (outc=>Float)))
+
+def conv2d (x:inc=>(Fin h)=>(Fin w)=>Float)
+           (kernel:outc=>inc=>(Fin kh)=>(Fin kw)=>Float) :
+     outc=>(Fin h)=>(Fin w)=>Float =
+     for o. for i.for j.
+         (i', j') = (ordinal i, ordinal j)
+         case (i' + kh) < h && (j' + kw) < w of
+          True ->
+              sum for ki:(Fin kh). sum for kj:(Fin kw).
+                  (di, dj) = (unsafeFromOrdinal (Fin h) (i' + (ordinal ki)),
+                              unsafeFromOrdinal (Fin w) (j' + (ordinal kj)))
+                  sum for inp. x.inp.di.dj * kernel.o.inp.ki.kj
+          False -> zero
+
+'TODO -> Figure out how to hide the h and w
+
+def cnn (h:Int) (w:Int) (inc:Type) (outc:Type) (kw:Int) (kh:Int) :
```

Comment:
Have you tried implicit parameters?
```suggestion
'TODO -> Figure out how to hide the h and w

def cnn (h:Int) ?-> (w:Int) ?-> (inc:Type) (outc:Type) (kw:Int) (kh:Int) :
```

Reply:
Amazing! For some reason I thought I had to do this on the type, which didn't work. But doing it on the function works great as long as the outerlayer specifies the size. 

https://github.com/google-research/dex-lang/pull/456/files#diff-bcced3748a9b282a26c9bd64cd2a907277e61fbe272d8a7f4854c3f8c6fb4129R156

Reply:
```haskell
lenet =
  H1 = Fin 4
  H2 = Fin 8
  H3 = Fin 64
  ncnn1 = cnn (Fin 1) H1 3 3
  ncnn2 = cnn H1 H2 3 3
  Pooled = (H2 & Fin 7 & Fin 7)
  ndense1 = dense Pooled H3
  ndense2 = dense H3 Class
  {
    forward = (\ (cnn1, cnn2, dense1, dense2) inp.
         x:Image = inp
         x1' = forward ncnn1 cnn1 x
         x1 = for i j k. relu x1'.i.j.k
         x2' = forward ncnn2 cnn2 x1
         x2 = for i j k. relu x2'.i.j.k
         x3 : (H2 => Fin 7 => Fin 7 => Float) = for c. meanpool (Fin 4) (Fin 4) x2.c
         x4' = forward ndense1 dense1 for (i,j,k). x3.i.j.k
         x4 = for i. relu x4'.i
         logsoftmax $ forward ndense2 dense2 x4),
    init = (\key.
         [k1, k2, k3, k4] = splitKey key
         (init ncnn1 k1, init ncnn2 k2,
         init ndense1 k3, init ndense2 k4))
  }
```

Mmm, satisfying af. Reminds me of [Hasktorch](https://github.com/hasktorch/hasktorch).